module-1/homework/SharedPointer/src/shared_ptr/shared_ptr.h

```diff
@@ -3,106 +3,294 @@
+    block_->IncreaseShared();
+}
+
+template<typename T>
+SharedPtr<T>::SharedPtr(SharedPtr&& other) noexcept : pointer_(other.pointer_), block_(other.block_) {
+    other.block_ = nullptr;
+    other.pointer_ = nullptr;
+}
+
+template<typename T>
+SharedPtr<T>& SharedPtr<T>::operator=(const SharedPtr& r) noexcept {
+    SharedPtr<T>(r).Swap(*this);
+    return *this;
+}
+
+template<typename T>
+SharedPtr<T>& SharedPtr<T>::operator=(SharedPtr&& r) noexcept {
+    SharedPtr<T>(std::move(r)).Swap(*this);
+    return *this;
+}
+
+template<typename T>
+void SharedPtr<T>::Reset() noexcept {
+    SharedPtr<T>().Swap(*this);
+}
+
+template<typename T>
+template<typename Y>
+void SharedPtr<T>::Reset(Y* p) noexcept {
+    SharedPtr<T>(p).Swap(*this);
+}
+
+template<typename T>
+template<typename Y, typename Deleter>
+void SharedPtr<T>::Reset(Y* p, Deleter deleter) noexcept {
+    SharedPtr<T>(p, deleter).Swap(*this);
+}
+
+template<typename T>
+void SharedPtr<T>::Swap(SharedPtr& other) noexcept {
+    std::swap(other.block_, block_);
+    std::swap(other.pointer_, pointer_);
+}
+
+template<typename T>
+T* SharedPtr<T>::Get() const noexcept {
+    return pointer_;
+}
+
+template<typename T>
+int64_t SharedPtr<T>::UseCount() const noexcept {
+    return block_ != nullptr ? block_->GetShared() : 0;
+}
+
+template<typename T>
+T& SharedPtr<T>::operator*() const noexcept {
+    return *pointer_;
+}
+
+template<typename T>
+T* SharedPtr<T>::operator->() const noexcept {
+    return Get();
+}
+
+template<typename T>
+typename SharedPtr<T>::element_type& SharedPtr<T>::operator[](std::ptrdiff_t idx) const {
+    return pointer_ != nullptr ? pointer_[idx] : nullptr;
+}
+
+template<typename T>
+SharedPtr<T>::operator bool() const noexcept {
+    return pointer_ != nullptr;
+}
 // SharedPtr
 
 // WeakPtr
-template <typename T>
+template<typename T>
 class WeakPtr {
 
 public:
-    using element_type =     // Your code goes here...
+    using element_type = T;
 
     // Special-member functions
     constexpr WeakPtr() noexcept = default;
-    template <typename Y>
+
+    template<typename Y>
     explicit WeakPtr(const SharedPtr<Y>& other);
+
     WeakPtr(const WeakPtr& other) noexcept;
+
     WeakPtr(WeakPtr&& other) noexcept;
-    template <typename Y>
+
+    template<typename Y>
     WeakPtr& operator=(const SharedPtr<Y>& other);
+
     WeakPtr& operator=(const WeakPtr& other) noexcept;
+
     WeakPtr& operator=(WeakPtr&& other) noexcept;
 
     ~WeakPtr();
 
     // Modifiers
     void Reset() noexcept;
+
     void Swap(WeakPtr<T>& other) noexcept;
 
     // Observers
-    bool Expired() noexcept;
+    bool Expired() const noexcept;
+
     SharedPtr<T> Lock() const noexcept;
 
-    template <typename U>
-    friend class SharedPtr;
+    template<typename U>
+    friend
+    class SharedPtr;
 
 public:
-    // Your code goes here...
+    SharedWeakCount* block_ = nullptr;
```

Comment:
[ES.23](Prefer the {}- initializer syntax)(https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-in-class-initializer)

```
SharedWeakCount* block_{nullptr};
```